aten/src/ATen/core/ivalue.h

```diff
@@ -207,14 +207,25 @@ struct CAFFE2_API IValue final {
   /// @private [doxygen private]
   c10::intrusive_ptr<caffe2::Blob> toBlob() const &;
 
-  // Capsule
-  IValue(intrusive_ptr<torch::CustomClassHolder> blob);
+  // Capsule. Capsule is an internal implementation detail
+  // of custom C++ classes. No new callsites of these APIs should
+  // be introduced.
+  static inline IValue make_capsule(intrusive_ptr<torch::CustomClassHolder> blob);
   bool isCapsule() const {
     return Tag::Capsule == tag;
   }
   c10::intrusive_ptr<torch::CustomClassHolder> toCapsule() &&;
   c10::intrusive_ptr<torch::CustomClassHolder> toCapsule() const &;
 
+  // Custom C++ classes
+  template <typename T, std::enable_if_t<std::is_base_of<torch::CustomClassHolder, T>::value, int> = 0>
+  IValue(intrusive_ptr<T> custom_class);
+  bool isCustomClass() const;
+  template <typename T>
+  c10::intrusive_ptr<T> toCustomClass() &&;
+  template <typename T>
+  c10::intrusive_ptr<T> toCustomClass() const &;
+
```

Comment:
Normally when we add a specific `toX` we also add to the templated `.to<X>`. Does `.to<MyCustomClass>()` work?

Reply:
it ends up being `.to<c10::intrusive_ptr<MyCustomClass>>()`. Maybe to normalize it we should kill the need to put `intrusive_ptr` there, I can look into it